crates/noirc_frontend/src/parser/parser.rs

```diff
@@ -653,8 +653,26 @@ where
 }
 
 fn expression() -> impl ExprParser {
-    recursive(|expr| expression_with_precedence(Precedence::Lowest, expr, false))
-        .labelled("expression")
+    recursive(|expr| {
+        choice((
+            return_expression(expr.clone()),
+            expression_with_precedence(Precedence::Lowest, expr, false),
+        ))
+    })
+    .labelled("expression")
+}
+
+fn return_expression<'a, P>(expr_parser: P) -> impl NoirParser<Expression> + 'a
+where
+    P: ExprParser + 'a,
+{
+    keyword(Keyword::Return)
+        .then(expr_parser)
```

Comment:
Lets use `ignore_then_commit(return, expr)` here since once we see a return we know we shouldn't fallback to another parser if we fail to parse an expr after it

Reply:
@jfecher Ok. Also, the expr after `return` should be optional with `.or_not()`.